File: src/IsValid.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class IsValid {
    /*
    *
    * 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。
    * 有效字符串需满足：
    * 左括号必须用相同类型的右括号闭合。
    * 左括号必须以正确的顺序闭合。
    * 注意空字符串可被认为是有效字符串。
    *
    * */
    public boolean isValid(String s) {
        if (s.length() == 0) {
            return true;
        }
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> symbolMap = generateSymbolMap();
        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);
            if (currentChar == '(' || currentChar == '{' || currentChar == '[') {
                stack.push(currentChar);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                Character pop = stack.pop();
                if (currentChar != symbolMap.get(pop)) {
                    return false;
                }
            }
        }
        if (stack.isEmpty()) {
            return true;
        } else {
            return false;
        }
    }

    private Map generateSymbolMap() {
        Map<Character, Character> map = new HashMap<>();
        map.put('(', ')');
        map.put('[', ']');
        map.put('{', '}');
        return map;
    }
}

File: src/stack/IsValid.kt
package stack

import java.util.*

class IsValid {
    fun isValid(s: String): Boolean {
        var stack = Stack<Char>()
        for (char in s) {
            if (stack.isNotEmpty()) {
                val peek = stack.peek()
                if (isValidStr(char, peek)) {
                    stack.pop()
                } else {
                    stack.push(char)
                }
            } else {
                stack.push(char)
            }
        }
        return stack.isEmpty()
    }

    fun isValidStr(s1: Char, s2: Char): Boolean {
        when (s1) {
            '(' -> return s2 == ')'
            '{' -> return s2 == '}'
            '[' -> return s2 == ']'
        }
        return false
    }
}